editor/src/clj/editor/scene.clj

```diff
@@ -1136,12 +1136,12 @@
           new-updatable-states (if (seq active-updatables)
                                  (profiler/profile "updatables" -1 (update-updatables updatable-states play-mode active-updatables))
                                  updatable-states)
-          renderables-invalidate-counter (g/invalidate-counter view-id :all-renderables evaluation-context)
-          last-renderables-invalidate-counter (ui/user-data image-view ::last-renderables-invalidate-counter)
+          renderables (g/node-value view-id :all-renderables evaluation-context)
+          last-renderables (ui/user-data image-view ::last-renderables)
           last-frame-version (ui/user-data image-view ::last-frame-version)
           frame-version (cond-> (or last-frame-version 0)
-                          (or (nil? last-renderables-invalidate-counter)
-                              (not= last-renderables-invalidate-counter renderables-invalidate-counter)
+                          (or (nil? last-renderables)
+                              (not= last-renderables renderables)
```

Comment:
We should probably use `identical?` here instead of equality test? I don't think we want to do a structural comparison, since these can be rather large structures.

Reply:
I profiled frame render time, here is the chart (not that Y-axis is logarithmic, not linear):

<img width="1109" alt="Screenshot 2022-05-11 at 11 35 11" src="https://user-images.githubusercontent.com/732554/167818773-8267cede-79a2-4893-b24c-f096e62a6463.png">
Y-axis shows spent time (ms), and X-axis is time.

blue line — `frame` — is frame rendering time in `editor.scene/update-image-view!`, i.e. a time spent rendering the frame; I added there `identical-eq` and `structural-eq` to see how identical and structural equality checks matter.

2 peaks on the graph appear when I move the object, and everything else is idle re-rendering. While it's true that a structural equality check is costlier, I don't think it matters since the difference is so tiny and happens on the scale of microseconds, while rendering occurs on the scale of multiple milliseconds.

Reply:
I'm not worried about the case where they do match - in that case `identical?` won't be any faster than the equality check. But the worst-case when they do not match is potentially really expensive to find with an equality check.

Consider the case where the position of a leaf-level sprite is modified in a deeply nested hierarchy. Using equality check, we would have to traverse the entire scene tree in order to find the difference. But in fact this is entirely unnecessary since we know for a fact that the scene has changed whenever the cached output produces a result that is not identical to the previous frame.

Reply:
Yes, your reasoning is correct, and charts show that: when outputs change, even though identity checks sometimes have higher peaks, structural checks generally perform worse. 

My concern is that generally, it doesn't matter because the difference is on the scale of microseconds in a process that takes milliseconds anyway, and it requires us to use `:cached` on `all-renderables` output to benefit from this optimization. I.e. this is an optimization that does not yield significant benefit while making the system a bit more rigid. Do you think it's worth it?

Reply:
Well, `all-renderables` is already cached in `SceneView`, so we'd only be adding caching to `PreviewView`, which shouldn't be a big deal as there typically will only be zero or one of those around in the view graph. I think adding `:cached` to a rare output is a small price to pay for making stuff more efficient. Every little helps, and here we have a clear situation where we know we can skip a bunch of work with a very small change.

Reply:
Sounds right!

Reply:
Although I realise now that maybe this approach of relying on caching is fundamentally incompatible with the ideas for "cost-based output caching" we talked a bit about earlier. Perhaps it is safer to use the equality check after all... I wonder if we have other situations like this. 🤔 I guess any use of `identical?` should be viewed with suspicion if we decide to try that idea out.

Reply:
Yes, but I also think some outputs, like `all-renderables`, probably should be always cached anyway

Reply:
Anyway. I saw that you pushed a commit to use `identical?`. Don't forget to add :cached to the `PreviewView` output if you decide to go through with it. 😅